ayman-art/BugByte
# Bugbyte 90 community service

### Description

#### **CommunityService**:
- A new service class `CommunityService` has been added to handle business logic for managing communities in the system.
- It provides functionality to interact with the `CommunityRepository` and perform operations such as creating, updating, fetching, and deleting communities.
- The service ensures that community data is efficiently managed and cached to optimize performance by reducing the number of database queries.

#### **Caching Mechanism**:
- `CommunityService` uses an in-memory cache to store recently accessed community data. 
- This cache (`communityPool`) is a `LinkedHashMap` with a capacity limit to store up to 100 communities. 
- When the cache exceeds the capacity, it automatically removes the least recently accessed community to free up space.

#### **Community Moderation**:
- The service allows setting and removing moderators for each community.
- It provides methods to assign and remove moderator roles for users within a specific community.
  
#### **CommunityServiceTest**:
- Unit tests for `CommunityService` are implemented in the `CommunityServiceTest` class to ensure that all service methods function correctly.
- The tests verify the following:
  - **CRUD operations**: Ensuring that communities are created, fetched, updated, and deleted as expected.
  - **Moderator Management**: Verifying that moderators can be added and removed correctly.
- The tests use **JUnit** and **Mockito** to mock dependencies and validate the logic of the service without interacting with the actual database.

### How to Test:
- To test the functionality of `CommunityService`, run the unit tests in the `CommunityServiceTest` class.
  - These tests validate the correctness of CRUD operations and moderator management.
  - They ensure that the service interacts properly with the `CommunityRepository` and correctly handles business logic.
